### PR TITLE
OCPBUGS-4101: Do not allow empty system reserved values

### PR DESCRIPTION
**Pull request by harche**
Signed-off-by: Harshal Patil <harpatil@redhat.com>

Fixes: https://issues.redhat.com/browse/OCPBUGS-4101

**- What I did**
Make sure none of the system reserved values end up empty and make kubelet crash during startup. 

**- How to verify it**
```
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: system1-reserved-config
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  kubeletConfig:
    systemReserved:
      cpu: 500m
      memory: 1500Mi
      ephemeral-storage: "" 
```
Even after deploying that MC, the kubelet should start up and come up with default value of the ephemeral storage for system reserved (1Gi)

**- Description for the changelog**

Force default values of system reserved in case empty values were supplied.

### CODE REVIEW

**Review by yuqi-zhang — APPROVED** [review]
lgtm, do you want to have Rio from the MCO QE test this? Or is there someone from the node team's side that can help verify?

**Comment by harche** [comment]
> lgtm, do you want to have Rio from the MCO QE test this? Or is there someone from the node team's side that can help verify?

From the node team I have verified it, so it would be better if MCO QE also verifies it. Thanks.

**Comment by yuqi-zhang** [comment]
@rioliu-rh is this something you would like to verify pre-merge?

**Comment by harche** [comment]
/hold 

They are manually overriding files internally used for auto node sizing using machine config. Then even if we handle this in MCO, their upgrade will still fail. I am not sure at this point if this should be even supported. Manually overriding internally used files are difficult to protect for external interference. 

https://issues.redhat.com/browse/OCPBUGS-4101?focusedCommentId=21354604&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21354604

**Comment by harche** [comment]
I will spend some more time to think if we can still fix it somehow for them.

**Comment by rioliu-rh** [comment]
> @rioliu-rh is this something you would like to verify pre-merge?

@yuqi-zhang let me try to verify it with cluster-bot image

**Comment by rioliu-rh** [comment]
tried to verify this fix , but it does not work for me

apply node sizing mc and kubelet config based on 4.11.16 successfully 

```
cat node-sizing-enabled-for-worker.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  name: 99-worker-node-sizing
  labels:
    machineconfiguration.openshift.io/role: worker
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      files:
      - contents:
          compression: ""
          source: data:text/plain;charset=utf-8;base64,Tk9ERV9TSVpJTkdfRU5BQkxFRD1mYWxzZQpTWVNURU1fUkVTRVJWRURfTUVNT1JZPTNHaQpTWVNURU1fUkVTRVJWRURfQ1BVPTEwMDBtCg==
        mode: 420
        overwrite: true
        path: /etc/node-sizing-enabled.env

oc create -f node-sizing-enabled-for-worker.yaml
machineconfig.machineconfiguration.openshift.io/99-worker-node-sizing created

oc debug node/ci-ln-smybk7b-72292-x29cs-worker-a-gdhlg -- chroot /host cat /etc/node-sizing-enabled.env
Temporary namespace openshift-debug-kd6mt is created for debugging node...
Starting pod/ci-ln-smybk7b-72292-x29cs-worker-a-gdhlg-debug ...
To use host binaries, run `chroot /host`
NODE_SIZING_ENABLED=false
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=1000m

curl -sL https://raw.githubusercontent.com/openshift/managed-cluster-config/d4522b5245ae9ef97d579ec2e661b12f7d256dfd/deploy/kubelet-config/01-kubelet-config.yaml > kubelet-config.yaml

cat kubelet-config.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: custom-kubelet
spec:
  machineConfigPoolSelector:
    matchExpressions:
    - key: machineconfiguration.openshift.io/mco-built-in
      operator: Exists
  autoSizingReserved: true

oc create -f kubelet-config.yaml
kubeletconfig.machineconfiguration.openshift.io/custom-kubelet created

oc debug node/ci-ln-smybk7b-72292-x29cs-worker-a-gdhlg -- chroot /host cat /etc/kubernetes/kubelet.conf | jq .systemReserved
Temporary namespace openshift-debug-qmr9b is created for debugging node...
Starting pod/ci-ln-smybk7b-72292-x29cs-worker-a-gdhlg-debug ...
To use host binaries, run `chroot /host`
{
  "ephemeral-storage": "1Gi"
}

```
upgrade the cluster to cluster-bot built image. `registry.build05.ci.openshift.org/ci-ln-qhxbptb/release:latest`

```
get auth token from https://console-openshift-console.apps.build05.l9oh.p1.openshiftapps.com/k8s/cluster/projects/ci-ln-qhxbptb and reset the pull-secret

oc adm upgrade --to-image registry.build05.ci.openshift.org/ci-ln-qhxbptb/release:latest --force --allow-explicit-upgrade

Upgradeable=False

  Reason: AdminAckRequired
  Message: Kubernetes 1.25 and therefore OpenShift 4.12 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6955381 for details and instructions.


oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.11-kube-1.25-api-removals-in-4.12":"true"}}' --type=merge
configmap/admin-acks patched

```
nodes are not ready
```
node
NAME                                       STATUS                        ROLES    AGE     VERSION
ci-ln-smybk7b-72292-x29cs-master-0         Ready                         master   3h32m   v1.25.2+5533733
ci-ln-smybk7b-72292-x29cs-master-1         NotReady,SchedulingDisabled   master   3h32m   v1.24.6+5157800
ci-ln-smybk7b-72292-x29cs-master-2         Ready                         master   3h33m   v1.24.6+5157800
ci-ln-smybk7b-72292-x29cs-worker-a-gdhlg   NotReady,SchedulingDisabled   worker   3h25m   v1.24.6+5157800
ci-ln-smybk7b-72292-x29cs-worker-b-g2hp8   Ready                         worker   3h24m   v1.24.6+5157800
ci-ln-smybk7b-72292-x29cs-worker-c-t5st4   Ready
```

**Comment by harche** [comment]
@rioliu-rh Thanks for testing, however I am not sure if the node sizing MC should be applied in the first place. That MC is modifying the files used by auto node sizing directly. Also, the system reserved values it is trying to set should be set using kubelet config. That's the only supported way.

**Comment by rioliu-rh** [comment]
> @rioliu-rh Thanks for testing, however I am not sure if the node sizing MC should be applied in the first place. That MC is modifying the files used by auto node sizing directly. Also, the system reserved values it is trying to set should be set using kubelet config. That's the only supported way.

So, I just need to apply below mc
```
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: system1-reserved-config
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  kubeletConfig:
    systemReserved:
      cpu: 500m
      memory: 1500Mi
      ephemeral-storage: "" 
```
then upgrade the cluster to the image which contains this fix

**Comment by rioliu-rh** [comment]
verified with following steps. @harche 

apply following mc
```
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: system-reserved-config
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  kubeletConfig:
    systemReserved:
      cpu: 500m
      memory: 1500Mi
      ephemeral-storage: ""

apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: enable-auto-sizing-reserved
spec:
  machineConfigPoolSelector:
    matchExpressions:
    - key: machineconfiguration.openshift.io/mco-built-in
      operator: Exists
  autoSizingReserved: true

```
mcp update is completed successfully

upgrade cluster to the image contains this pr registry.build05.ci.openshift.org/ci-ln-80tcgqk/release:latest

upgrade is success
```
oc adm upgrade
Cluster version is 4.12.0-0.ci.test-2022-12-06-083703-ci-ln-80tcgqk-latest
```

check file content of `/etc/node-sizing-enabled.env` on worker node

```
oc debug node/rioliu-1206b-ldgpl-worker-a-7cr96.c.openshift-qe.internal -- chroot /host cat /etc/node-sizing-enabled.env
Temporary namespace openshift-debug-r2gl7 is created for debugging node...
Starting pod/rioliu-1206b-ldgpl-worker-a-7cr96copenshift-qeinternal-debug ...
To use host binaries, run `chroot /host`
NODE_SIZING_ENABLED=true
SYSTEM_RESERVED_MEMORY=1Gi
SYSTEM_RESERVED_CPU=500m
SYSTEM_RESERVED_ES=1Gi
```

check file content of `kubelet.conf` on worker node

```
oc debug node/rioliu-1206b-ldgpl-worker-a-7cr96.c.openshift-qe.internal -- chroot /host cat /etc/kubernetes/kubelet.conf
Temporary namespace openshift-debug-7wr89 is created for debugging node...
Starting pod/rioliu-1206b-ldgpl-worker-a-7cr96copenshift-qeinternal-debug ...
To use host binaries, run `chroot /host`
{
  "kind": "KubeletConfiguration",
  "apiVersion": "kubelet.config.k8s.io/v1beta1",
  "staticPodPath": "/etc/kubernetes/manifests",
  "syncFrequency": "0s",
  "fileCheckFrequency": "0s",
  "httpCheckFrequency": "0s",
  "tlsCipherSuites": [
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
  ],
  "tlsMinVersion": "VersionTLS12",
  "rotateCertificates": true,
  "serverTLSBootstrap": true,
  "authentication": {
    "x509": {
      "clientCAFile": "/etc/kubernetes/kubelet-ca.crt"
    },
    "webhook": {
      "cacheTTL": "0s"
    },
    "anonymous": {
      "enabled": false
    }
  },
  "authorization": {
    "webhook": {
      "cacheAuthorizedTTL": "0s",
      "cacheUnauthorizedTTL": "0s"
    }
  },
  "clusterDomain": "cluster.local",
  "clusterDNS": [
    "172.30.0.10"
  ],
  "streamingConnectionIdleTimeout": "0s",
  "nodeStatusUpdateFrequency": "0s",
  "nodeStatusReportFrequency": "0s",
  "imageMinimumGCAge": "0s",
  "volumeStatsAggPeriod": "0s",
  "systemCgroups": "/system.slice",
  "cgroupRoot": "/",
  "cgroupDriver": "systemd",
  "cpuManagerReconcilePeriod": "0s",
  "runtimeRequestTimeout": "0s",
  "maxPods": 250,
  "podPidsLimit": 4096,
  "kubeAPIQPS": 50,
  "kubeAPIBurst": 100,
  "serializeImagePulls": false,
  "evictionPressureTransitionPeriod": "0s",
  "featureGates": {
    "APIPriorityAndFairness": true,
    "CSIMigrationAzureFile": false,
    "CSIMigrationvSphere": false,
    "DownwardAPIHugePages": true,
    "RotateKubeletServerCertificate": true
  },
  "memorySwap": {},
  "containerLogMaxSize": "50Mi",
  "logging": {
    "flushFrequency": 0,
    "verbosity": 0,
    "options": {
      "json": {
        "infoBufferSize": "0"
      }
    }
  },
  "shutdownGracePeriod": "0s",
  "shutdownGracePeriodCriticalPods": "0s"
}
```

**Comment by harche** [comment]
@rioliu-rh I made the final script more resilient to the unexpected input. 

```bash
sh-4.4#  /usr/local/sbin/dynamic-system-reserved-calc.sh true 
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=0.08
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
sh-4.4# 
sh-4.4#  /usr/local/sbin/dynamic-system-reserved-calc.sh false
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=1Gi
SYSTEM_RESERVED_CPU=500m
SYSTEM_RESERVED_ES=1Gi
sh-4.4#                          
sh-4.4# 
sh-4.4#  /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi 1000m 1.5Gi
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=1000m
SYSTEM_RESERVED_ES=1.5Gi
sh-4.4# 
sh-4.4#  /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi 1000m      
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=1000m
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
sh-4.4#  /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi       
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=500m
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 

```

**Comment by rioliu-rh** [comment]
@harche is it expected, if yes, I will add  label `qe-approve`

**Comment by harche** [comment]
> @harche is it expected, if yes, I will add label `qe-approved` cc: @yuqi-zhang

Sounds good. Please add the `qe-approved` label whenever you are ready. 

 I also did additional testing with the actual script that was generating empty value for `SYSTEM_RESERVED_ES` variable.  With that fix, the script should never output any of system reserved values empty irrespective of the input. 

```
sh-4.4# /usr/local/sbin/dynamic-system-reserved-calc.sh true
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=0.08
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
sh-4.4# 
sh-4.4# /usr/local/sbin/dynamic-system-reserved-calc.sh false
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=1Gi
SYSTEM_RESERVED_CPU=500m
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
sh-4.4# 
sh-4.4# /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi 1000m 1.5Gi
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=1000m
SYSTEM_RESERVED_ES=1.5Gi
sh-4.4# 
sh-4.4# /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi 1000m 
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=1000m
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
sh-4.4# /usr/local/sbin/dynamic-system-reserved-calc.sh false 3Gi 
sh-4.4# cat /etc/node-sizing.env 
SYSTEM_RESERVED_MEMORY=3Gi
SYSTEM_RESERVED_CPU=500m
SYSTEM_RESERVED_ES=1Gi
sh-4.4# 
 ```

**Comment by yuqi-zhang** [comment]
I am +1 and can apply lgtm once we are satisfied

**Pull request merged**
